model/value.go

```diff
@@ -221,80 +187,74 @@ func (s Samples) Equal(o Samples) bool {
 
 // SampleStream is a stream of Values belonging to an attached COWMetric.
 type SampleStream struct {
-	Metric Metric       `json:"metric"`
-	Values []SamplePair `json:"values"`
+	Metric     Metric                `json:"metric"`
+	Values     []SamplePair          `json:"values"`
+	Histograms []SampleHistogramPair `json:"histograms"`
+	Type       SampleType
```

Comment:
I fairly certain that we have to support SampleStreams of mixed type. (The Prometheus query API can definitely return a series that has both.) I think you can just get rid of the `Type` field. The usual case of a pure histogram SampleStream or a pure float SampleStream would then simply be represented by the corresponding other slice being empty. A mixed SampleStream would have both slices populated. (And that matches the semantics of the JSON response in the query API, IIRC.)

Reply:
Oh I didn't know that, I've only observed JSON responses with either values or histograms, not both. In what cases would it have both? How can we instrument a server that sends mixed type metrics, or write a query that returns both? It'll help to know what these cases mean so that we can handle the corresponding logic properly in Mimir (we've been assuming only one type per time series / sample stream so far, and that if we somehow receive both, it's a mistake and we'll ignore the floats). Would the mixed type mean that the promql.Points have both V and H defined?

Reply:
> Would the mixed type mean that the promql.Points have both V and H defined?

No. With mixed type we mean that in a stream of samples a sample can be float, and the next sample can be a histogram for example. It is possible when someone is migrating from old histogram (with float series) to new histograms, or conflicting names for instrumentation (a user errors).

Example of what we mean, for the above struct (basically both samples can co-exist in a stream).
```
Metric: {foo="bar"}
Values: [ (100, 1.0), (200, 2.2), (300, 3.3) ]
Histograms: [ (150, h1), (250, h2) ]
```

Reply:
Yeah, it's definitely expected to be rare, but it is possible, and thus we have to handle it.